### PR TITLE
feat: odp datafile parsing and audience evaluation

### PR DESCRIPTION
**Pull request by andrewleap-optimizely**
Summary
-------

- A new field `integrations` and new audience type `odp.audiences` will be added to the datafile schema for ODP integration.

-  parse the odp values (`publicKey` and `host`)

-  accept datafile with and without the `integrations` section or empty section as well.

-  support a new audience with the `odp.audiences` and `qualified` match.

-  support any (allowed) combination of audiences, including ODP and other existing audiences.

Test plan
---------
Added/extended tests in:
- test_optimizely_config.py
- test_validator.py
- test_audience.py
- test_condition.py
- test_user_context.py

### CODE REVIEW

File: tests/test_optimizely_config.py [hunk 1]
```diff
@@ -1800,3 +1802,46 @@ def test_get_delivery_rules(self):
                 self.assertIsInstance(delivery_rule, optimizely_config.OptimizelyExperiment)
 
         self.assertEqual(expected_features_map_dict, actual_features_map_dict)
+
```
Comment (jaeopt) [inline tests/test_optimizely_config.py:1805]: we can remove these tests for optimizely_config

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks great! A few small changes suggested. 
More tests for datafile "integrations" processing including invalid format, missing, and empty will be good.
https://github.com/optimizely/swift-sdk/blob/jae/ats/Tests/OptimizelyTests-DataModel/IntegrationTests.swift

File: tests/test_optimizely_config.py (hunk 1 above)
Comment (jaeopt) [reply tests/test_optimizely_config.py:1805]: We can move these tests to project_config tests.

Comment (andrewleap-optimizely) [reply tests/test_optimizely_config.py:1805]: done

File: optimizely/helpers/condition.py [hunk 2]
```diff
@@ -55,21 +55,23 @@ class ConditionMatchTypes:
...
     """ Class encapsulating methods to be used in audience leaf condition evaluation. """
 
-    CUSTOM_ATTRIBUTE_CONDITION_TYPE: Final = 'custom_attribute'
+    CONDITION_TYPES: Final = ('custom_attribute', 'third_party_dimension')
 
     def __init__(
         self,
         condition_data: list[str | list[str]],
-        attributes: Optional[optimizely_user_context.UserAttributes],
+        user_context: optimizely_user_context.OptimizelyUserContext,
```
[the remark this answers is not in the document]
Comment (jaeopt) [reply optimizely/helpers/condition.py:69]: Most of them look like non-public type/methods, but if you do not feel comfortable about possible breaking changes, let's leave them as is.  We can try to add new ones with consistent names (either user or user_context).  Not 100% but a general rule is that any type starting with "Optimizely" are public.

**Review by Mat001 — CHANGES_REQUESTED** [review]

File: optimizely/helpers/condition.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewleap-optimizely) [reply optimizely/helpers/condition.py:69]: If they are considered non-public, I think it's fine, as long as we call it out in the release notes. Especially since it's actually only `get_variation_for_rollout` that is currently using `user`.

File: optimizely/optimizely_user_context.py [hunk 3]
```diff
@@ -95,6 +96,8 @@ def _clone(self) -> Optional[OptimizelyUserContext]:
         with self.lock:
             if self.forced_decisions_map:
                 user_context.forced_decisions_map = copy.deepcopy(self.forced_decisions_map)
+            if self.qualified_segments:
+                user_context.qualified_segments = self.qualified_segments.copy()
```
[the remark this answers is not in the document]
Comment (andrewleap-optimizely) [reply optimizely/optimizely_user_context.py]: added comments

**Review by Mat001 — APPROVED** [review]
lgtm

File: tests/helpers_tests/test_condition.py [hunk 4]
```diff
@@ -1206,15 +1206,15 @@ def test_invalid_semver__returns_None__when_semver_is_invalid(self):
...
+        self.user_context._qualified_segments = ['odp-segment-2']
         evaluator = condition_helper.CustomAttributeConditionEvaluator(
             qualified_condition_list, self.user_context, self.mock_client_logger,
         )
 
         self.assertStrictTrue(evaluator.evaluate(0))
 
     def test_qualified__returns_false__when_user_is_not_qualified(self, ):
-        self.user_context.qualified_segments = ['odp-segment-1']
+        self.user_context._qualified_segments = ['odp-segment-1']
```
Comment (jaeopt) [inline tests/helpers_tests/test_condition.py:1217]: Don't we need support for atomic getter for "qualified_segments"?  like "get_user_attributes()"

File: tests/test_config.py [hunk 5]
```diff
@@ -1024,6 +1026,64 @@ def test_to_datafile_from_bytes(self):
 
         self.assertEqual(expected_datafile, actual_datafile)
 
+    def test_datafile_with_audience_segments(self):
```
Comment (jaeopt) [inline tests/test_config.py:1029]:
```suggestion
    def test_datafile_with_integrations(self):
```

File: tests/test_config.py [hunk 6]
```diff
@@ -1024,6 +1026,64 @@ def test_to_datafile_from_bytes(self):
...
+        self.assertGreater(len(integrations), 0)
+        self.assertEqual(len(project_config.integrations), len(integrations))
+
+        integration = integrations[0]
+        self.assertEqual(project_config.host_for_odp, integration['host'])
+        self.assertEqual(project_config.public_key_for_odp, integration['publicKey'])
+
+        self.assertEqual(sorted(project_config.all_segments), ['odp-segment-1', 'odp-segment-2', 'odp-segment-3'])
+
+    def test_datafile_with_no_audience_segments(self):
```
Comment (jaeopt) [inline tests/test_config.py:1050]:
```suggestion
    def test_datafile_with_no_integrations(self):
```

File: tests/test_config.py [hunk 7]
```diff
@@ -1024,6 +1026,64 @@ def test_to_datafile_from_bytes(self):
...
+        opt_obj = optimizely.Optimizely(
+            json.dumps(config_dict_with_audience_segments)
+        )
+
+        project_config = opt_obj.config_manager.get_config()
+
+        self.assertIsInstance(project_config, ProjectConfig)
+        self.assertEqual(len(project_config.integrations), 0)
+
+    def test_datafile_with_audience_segments_missing_key(self):
```
Comment (jaeopt) [inline tests/test_config.py:1063]:
```suggestion
    def test_datafile_with_integrations_missing_key(self):
```

File: tests/test_config.py [hunk 8]
```diff
@@ -1024,6 +1026,64 @@ def test_to_datafile_from_bytes(self):
...
+        del config_dict_with_audience_segments['integrations'][0]['host']
+        opt_obj = optimizely.Optimizely(
+            json.dumps(config_dict_with_audience_segments)
+        )
+
+        project_config = opt_obj.config_manager.get_config()
+
+        self.assertIsNone(project_config)
+
+    def test_datafile_with_invalid_audience_segments(self):
```
Comment (jaeopt) [inline tests/test_config.py:1075]: Can we replace this test with "key" only integration (missing other fields), and it's accepted ok?
The only field in "integrations" is "key", while all others optional.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Most changes look good. I suggested a few more changes.
Not sure about the plan for atomic (public) getter/setter in optimizely_user_context. We can chat.

File: tests/helpers_tests/test_condition.py (hunk 4 above)
Comment (andrewleap-optimizely) [reply tests/helpers_tests/test_condition.py]: My bad, I thought we were going to implement those during a subsequent PR. I've added a getter and setter for qualified segments.

**Review by jaeopt — APPROVED** [review]
LGTM
